### PR TITLE
Add elementwise fusion on multiuse ops

### PR DESCRIPTION
**Pull request by pzread**
Add the option to enable elementwise fusion on multiuse ops.

The code is guarded by `iree-flow-enable-aggressive-fusion` and disabled by default. We need some fixes in the backends to make this work.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp [hunk 1]
```diff
@@ -57,127 +86,273 @@ static bool areFusableOps(MLIRContext *context, Operation *producerOp,
...
     registry.insert<AffineDialect, linalg::LinalgDialect, math::MathDialect>();
   }
+  FusionOfTensorOpsPass(bool fuseMultiUse, unsigned multiUseFusionIteration)
+      : fuseMultiUse(fuseMultiUse),
+        multiUseFusionIteration(multiUseFusionIteration) {}
+  FusionOfTensorOpsPass(const FusionOfTensorOpsPass &pass)
+      : fuseMultiUse(pass.fuseMultiUse),
+        multiUseFusionIteration(pass.multiUseFusionIteration) {}
+
+  LogicalResult initializeOptions(StringRef options) override {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp:201]: Huh, interesting. Are you sure this is not the other way round. IIUC the options added in tblgen are already members of the base class. You can use the values passed in the constructor to assign the values to the options, and then you can just leave the base class `initializeOptions` as is and not override them. Within the code you can use the option directly as a `bool` or `unsigned` value....

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 2]
```diff
@@ -84,6 +84,12 @@ static llvm::cl::opt<bool> clEnableLinalgDetensorize(
     llvm::cl::desc("Enable detensorizing linalg ops to operate on primitives"),
     llvm::cl::init(true));
 
+static llvm::cl::opt<bool> clEnableFusionOnMultiUseAndReduction(
+    "iree-flow-enable-fusion-on-multi-use-and-reduction",
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp:88]: Could we just call this `iree-flow-enable-aggressive-fusion`?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Nice! Overall looks great! Just a clarification about the pass options.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp (hunk 2 above)
Comment (pzread) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp]: Done.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp (hunk 1 above)
Comment (pzread) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp]: SG. Thx.

**Comment by MaheshRavishankar** [comment]
Hi @pzread is there an ASAN issue here? or did you already find it?

**Review by MaheshRavishankar — APPROVED** [review]
Looks good to me. Thanks!

**Comment by pzread** [comment]
> Hi @pzread is there an ASAN issue here? or did you already find it?

Found the issue... I only populated createFusionOfTensorOpsPass on `func::FuncOp` but not `IREE:Util::InitializerOp`

**Pull request merged**